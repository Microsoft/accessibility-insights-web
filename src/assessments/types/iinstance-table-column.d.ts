// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { IAssessmentInstanceRowData } from '../../DetailsView/components/assessment-instance-table';

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: IAssessmentInstanceRowData) => JSX.Element;
}
