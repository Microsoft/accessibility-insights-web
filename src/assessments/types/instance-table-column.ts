// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GeneratedAssessmentInstance } from 'common/types/store-data/assessment-result-data';

export interface AssessmentInstanceRowData<P = {}> {
    key: string;
    statusChoiceGroup: JSX.Element;
    visualizationButton?: JSX.Element;
    instance: GeneratedAssessmentInstance<P>;
}

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: AssessmentInstanceRowData) => JSX.Element;
}
