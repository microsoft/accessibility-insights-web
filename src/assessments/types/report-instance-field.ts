// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    ColumnValue,
    ColumnValueBag,
    isScalarColumnValue,
} from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { TestStepInstance } from 'common/types/store-data/assessment-result-data';

export type ReportInstanceField<Instance = Partial<TestStepInstance>> = {
    key: string;
    label: string;
    getValue: (instance: Instance) => ColumnValue;
};

export type ReportInstanceFieldMap = { [KEY in string]?: ReportInstanceField };

export type ReportInstanceFields = ReportInstanceField[];

type HasPropertyBag<PB> = { propertyBag: PB };

function fromPropertyBagField<PB>(
    label: string,
    key: keyof PB & string,
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        return i.propertyBag && i.propertyBag[key] && (i.propertyBag[key] as any).toString();
    }
    return { key, label, getValue };
}

function fromColumnValueBagField<PB extends ColumnValueBag>(
    label: string,
    key: keyof PB & string,
    defaultValue = '-',
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        const value = i.propertyBag && i.propertyBag[key];
        return isValid(value) ? value : defaultValue;
    }

    return { key, label, getValue };
}

function fromPropertyBagFunction<PB>(
    label: string,
    key: string,
    accessor: (bag: PB) => string | null,
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        return i.propertyBag && accessor(i.propertyBag);
    }
    return { key, label, getValue };
}

type CommonReportInstanceFieldKey = 'comment' | 'snippet' | 'path' | 'manualSnippet' | 'manualPath';
const common: { [key in CommonReportInstanceFieldKey]: ReportInstanceField } = {
    comment: { key: 'comment', label: 'Comment', getValue: i => i.description },
    snippet: { key: 'snippet', label: 'Snippet', getValue: i => i.html },
    path: { key: 'path', label: 'Path', getValue: i => i.target && i.target.join(', ') },
    manualSnippet: { key: 'manualSnippet', label: 'Code Snippet', getValue: i => i.html },
    manualPath: { key: 'manualPath', label: 'CSS Selector', getValue: i => i.selector },
};

function isValid(value: ColumnValue): ColumnValue {
    if (!value) {
        return false;
    }
    if (!isScalarColumnValue(value) && Object.keys(value).length === 0) {
        return false;
    }
    return true;
}

function fromColumns<PB extends ColumnValueBag>(
    cfg: PropertyBagColumnRendererConfig<PB>[],
): ReportInstanceField[] {
    return cfg.map(fromColumnConfig);

    function fromColumnConfig({
        propertyName,
        defaultValue,
        displayName,
    }: PropertyBagColumnRendererConfig<PB>): ReportInstanceField<HasPropertyBag<PB>> {
        function getValue(i: HasPropertyBag<PB>): ColumnValue {
            const value = i.propertyBag && i.propertyBag[propertyName];
            return isValid(value) ? value : defaultValue;
        }

        return {
            key: propertyName,
            label: displayName,
            getValue,
        };
    }
}

function fromSnippet(key: string, label: string): ReportInstanceField {
    return { key, label, getValue: i => i.html };
}

export const ReportInstanceField = {
    fromPropertyBagField,
    fromColumnValueBagField,
    fromPropertyBagFunction,
    fromColumns,
    fromSnippet,
    common,
};
