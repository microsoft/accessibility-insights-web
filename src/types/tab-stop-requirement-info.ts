// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { GuidanceLink } from 'common/types/store-data/guidance-links';

export const TabStopRequirementIds = [
    'keyboard-navigation',
    'keyboard-traps',
    'focus-indicator',
    'tab-order',
    'input-focus',
] as const;

export type TabStopRequirementId = typeof TabStopRequirementIds[number];

export type TabStopRequirementInfo = {
    [requirementId in TabStopRequirementId]: TabStopRequirementContent;
};

export type TabStopRequirementContent = {
    name: string;
    description: string;
    guidance: GuidanceLink[];
};
