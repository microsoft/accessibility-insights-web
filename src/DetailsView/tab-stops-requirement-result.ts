// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { TabStopRequirementId } from 'types/tab-stop-requirement-info';

export interface TabStopsRequirementResultInstance {
    id: string;
    description: string;
}
export interface TabStopsRequirementResult {
    id: TabStopRequirementId;
    description: string;
    name: string;
    instances: TabStopsRequirementResultInstance[];
    isExpanded: boolean;
}
