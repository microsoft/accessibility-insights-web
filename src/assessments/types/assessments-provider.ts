// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { DictionaryStringTo } from 'types/common-types';
import { Assessment } from './iassessment';
import { Requirement } from './requirement';

export interface AssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(visualizationType: VisualizationType): boolean;
    forType(visualizationType: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    forRequirementKey(key: string): Readonly<Assessment>;
    getStep(visualizationType: VisualizationType, key: string): Readonly<Requirement>;
    getStepMap(
        visualizationType: VisualizationType,
    ): Readonly<DictionaryStringTo<Readonly<Requirement>>>;
}
