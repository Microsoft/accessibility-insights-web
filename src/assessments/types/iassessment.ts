// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RequirementOrdering } from '../../common/assessment/requirement';
import { IAssesssmentVisualizationConfiguration } from '../../common/configs/visualization-configuration-factory';
import { AnyExtension } from '../../common/extensibility/extension-point';
import { IAssessmentData } from '../../common/types/store-data/iassessment-result-data';
import { VisualizationType } from '../../common/types/visualization-type';
import { ContentPageComponent } from '../../views/content/content-page';
import { TestStep } from './test-step';

interface BaseAssessment {
    key: string;
    type: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    steps: TestStep[];
    featureFlag?: { required?: string[] };
    executeAssessmentScanPolicy?: (scheduleScan: (step: string) => void, data: IAssessmentData) => void;
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<IAssesssmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => IAssesssmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
