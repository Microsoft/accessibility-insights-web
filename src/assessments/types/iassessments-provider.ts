// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from '../../common/types/visualization-type';
import { Assessment } from './iassessment';
import { TestStep } from './test-step';

// tslint:disable-next-line:interface-name
export interface IAssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(type: VisualizationType): boolean;
    forType(type: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    getStep(type: VisualizationType, key: string): Readonly<TestStep>;
    getStepMap(type: VisualizationType): Readonly<IDictionaryStringTo<Readonly<TestStep>>>;
}
