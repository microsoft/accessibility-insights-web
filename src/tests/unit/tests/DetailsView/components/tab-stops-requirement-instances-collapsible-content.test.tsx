// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DetailsList } from '@fluentui/react';
import { act, fireEvent, render } from '@testing-library/react';
import {
    TabStopsRequirementInstancesCollapsibleContent,
    TabStopsRequirementInstancesCollapsibleContentProps,
} from 'DetailsView/tab-stops-requirement-instances-collapsible-content';
import { TabStopsRequirementResultInstance } from 'DetailsView/tab-stops-requirement-result';
import * as React from 'react';

import { IMock, Mock, Times } from 'typemoq';
import { TabStopRequirementId } from 'types/tab-stop-requirement-info';
import {
    expectMockedComponentPropsToMatchSnapshots,
    getMockComponentClassPropsForCall,
    mockReactComponents,
    useOriginalReactElements,
} from '../../../mock-helpers/mock-module-helpers';
jest.mock('@fluentui/react');

describe('TabStopsRequirementInstancesCollapsibleContent', () => {
    mockReactComponents([DetailsList]);
    let onEditButtonClickedMock: IMock<
        (requirementId: TabStopRequirementId, instanceId: string, description: string) => void
    >;
    let onRemoveButtonClickedMock: IMock<
        (requirementId: TabStopRequirementId, instanceId: string) => void
    >;
    let props: TabStopsRequirementInstancesCollapsibleContentProps;
    let requirementResultInstanceStub: TabStopsRequirementResultInstance;

    beforeEach(() => {
        onEditButtonClickedMock =
            Mock.ofType<
                (
                    requirementId: TabStopRequirementId,
                    instanceId: string,
                    description: string,
                ) => void
            >();
        onRemoveButtonClickedMock =
            Mock.ofType<(requirementId: TabStopRequirementId, instanceId: string) => void>();

        props = {
            requirementId: 'keyboard-navigation',
            instances: [{ id: 'test-instance-id', description: 'test-description' }],
            onEditButtonClicked: onEditButtonClickedMock.object,
            onRemoveButtonClicked: onRemoveButtonClickedMock.object,
        };

        requirementResultInstanceStub = {
            id: 'test-instance-id',
            description: 'test requirement description',
        };
    });

    it('renders', () => {
        const renderResult = render(<TabStopsRequirementInstancesCollapsibleContent {...props} />);
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([DetailsList]);
    });

    test('renders captured instance details column', () => {
        render(<TabStopsRequirementInstancesCollapsibleContent {...props} />);
        const columns = getMockComponentClassPropsForCall(DetailsList).columns;
        expect(columns[0].onRender(requirementResultInstanceStub)).toMatchSnapshot();
    });

    test('renders captured instance icons column', () => {
        render(<TabStopsRequirementInstancesCollapsibleContent {...props} />);
        const columns = getMockComponentClassPropsForCall(DetailsList).columns;
        expect(columns[1].onRender(requirementResultInstanceStub)).toMatchSnapshot();
    });

    test('click events pass through as expected', async () => {
        onEditButtonClickedMock
            .setup(ebc => ebc('keyboard-navigation', 'test-instance-id', 'test-description'))
            .verifiable(Times.once());
        onRemoveButtonClickedMock
            .setup(rbc => rbc(props.requirementId, 'test-instance-id'))
            .verifiable(Times.once());
        useOriginalReactElements('@fluentui/react', ['DetailsList', 'Link']);

        const renderResult = render(
            <TabStopsRequirementInstancesCollapsibleContent
                requirementId={props.requirementId}
                instances={props.instances}
                onEditButtonClicked={onEditButtonClickedMock.object}
                onRemoveButtonClicked={onRemoveButtonClickedMock.object}
            />,
        );

        const buttons = renderResult.getAllByRole('button');
        act(() => {
            buttons.forEach(button => fireEvent.click(button));
        });

        onEditButtonClickedMock.verifyAll();
        onRemoveButtonClickedMock.verifyAll();
    });
});
