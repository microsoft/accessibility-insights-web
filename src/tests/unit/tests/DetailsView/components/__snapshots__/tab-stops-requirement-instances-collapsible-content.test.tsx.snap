// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabStopsRequirementInstancesCollapsibleContent renders 1`] = `
<DocumentFragment>
  <mock-styledwithviewportcomponent
    checkboxvisibility="2"
    columns="[object Object],[object Object]"
    constrainmode="1"
    items="[object Object]"
  />
</DocumentFragment>
`;

exports[`TabStopsRequirementInstancesCollapsibleContent renders captured instance details column 1`] = `
<AssessmentInstanceDetailsColumn
  background="#767676"
  headerText="Comment"
  textContent="test requirement description"
/>
`;

exports[`TabStopsRequirementInstancesCollapsibleContent renders captured instance icons column 1`] = `
<React.Fragment>
  <ForwardRef(StyledLinkBase)
    className="editButton"
    onClick={[Function]}
  >
    <Memo
      ariaLabel="edit instance"
      iconName="edit"
    />
  </ForwardRef(StyledLinkBase)>
  <ForwardRef(StyledLinkBase)
    className="removeButton"
    onClick={[Function]}
  >
    <Memo
      ariaLabel="delete instance"
      iconName="delete"
    />
  </ForwardRef(StyledLinkBase)>
</React.Fragment>
`;

exports[`TabStopsRequirementInstancesCollapsibleContent renders: StyledWithViewportComponent props 1`] = `
{
  "checkboxVisibility": 2,
  "columns": [
    {
      "columnActionsMode": 0,
      "fieldName": "description",
      "isResizable": true,
      "key": "failureDescription",
      "minWidth": 100,
      "name": "Failure description",
      "onRender": [Function],
    },
    {
      "columnActionsMode": 0,
      "fieldName": "instanceActionButtons",
      "isIconOnly": true,
      "isResizable": false,
      "key": "instanceActionButtons",
      "maxWidth": 124,
      "minWidth": 124,
      "name": "instance actions",
      "onRender": [Function],
    },
  ],
  "constrainMode": 1,
  "items": [
    {
      "description": "test-description",
      "id": "test-instance-id",
    },
  ],
  "onRenderDetailsHeader": [Function],
}
`;
